examples/rlhf/train_rlhf.py

```diff
@@ -0,0 +1,385 @@
+        return tensordict.update(td)
+
+
+def logprobs_of_labels(logits, labels):
+    """Log probabilities of the labels
+
+    These are calculated from the logits."""
+    logprobs = F.log_softmax(logits, dim=-1)
+    logprobs_labels = torch.gather(logprobs, dim=-1, index=labels.unsqueeze(-1))
+    return logprobs_labels.squeeze(-1)
+
+
+@torch.no_grad()
+def generate(model, batch, ref_model, max_new_tokens=50):
+    input_ids = batch.mask_label().input_ids
+
+    # move padding tokens to left pad
+    # huggingface models expect left padding for generation
+    input_ids = torch.stack(
+        [torch.roll(row, (row == EOS_TOKEN_ID).sum().item(), 0) for row in input_ids]
+    )
+
+    # generate and capture scores
+    generation_config = GenerationConfig(
+        pad_token_id=EOS_TOKEN_ID,
+        max_new_tokens=max_new_tokens,
+        return_dict_in_generate=True,
+        output_scores=True,
+    )
+    outputs = model.generate(
+        input_ids=input_ids,
+        attention_mask=(input_ids != EOS_TOKEN_ID).to(torch.int64),
+        generation_config=generation_config,
+    )
+    samples = outputs.sequences
+    # we'll insert generated tokens into a tensor prepopulated with padding tokens,
+    # thereby moving back to right padding for reward model
+    generated_shape = torch.Size(
+        [input_ids.shape[0], input_ids.shape[1] + max_new_tokens]
+    )
+    generated = (
+        torch.ones(generated_shape, dtype=input_ids.dtype, device=input_ids.device)
+        * EOS_TOKEN_ID
+    )
+    for i, sample in enumerate(samples):
+        mask = sample != EOS_TOKEN_ID
+        generated[i, : mask.sum()] = sample[mask]
+
+    # get the scores and normalise for log probabilities
+    scores = torch.stack(outputs.scores, 1)
+    log_probs_gen = F.pad(
+        F.log_softmax(scores, dim=-1).max(dim=-1).values,
+        (0, max_new_tokens - scores.shape[1]),
+        value=0,
+    )
+
+    # get the scores and normalise for log probabilities
+    attention_mask = (generated != EOS_TOKEN_ID).to(torch.int64)
+    logits = model(
+        input_ids=generated, attention_mask=attention_mask, return_dict=True
+    ).logits
+    logprobs = logprobs_of_labels(logits[:, :-1], generated[:, 1:])
+    ref_logits = ref_model(
+        input_ids=generated.to(ref_model.device),
+        attention_mask=attention_mask.to(ref_model.device),
+        return_dict=True,
+    ).logits.to(logits.device)
+    ref_logprobs = logprobs_of_labels(ref_logits[:, :-1], generated[:, 1:])
+    log_ratio = (logprobs - ref_logprobs) * attention_mask[:, :-1]
+    log_ratio = torch.stack(
+        [
+            row[rindex - 1 : rindex + max_new_tokens - 1]
+            for row, rindex in zip(log_ratio, batch.prompt_rindex)
+        ],
+        dim=0,
+    )
+    return generated, log_probs_gen, log_ratio
+
+
+@torch.no_grad()
+def create_rollout_td(
+    batch, generated, reward_model, log_probs, log_ratio, max_new_tokens=50, kl_coef=0.1
+):
```

Comment:
worth porting to repo?

Reply:
See comment on `log_probs_of_labels` above.